hathor/websocket/iterators.py

```diff
@@ -0,0 +1,156 @@
+# Copyright 2024 Hathor Labs
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from collections import deque
+from collections.abc import AsyncIterable
+from dataclasses import dataclass
+from typing import AsyncIterator
+
+from twisted.internet.defer import Deferred
+
+from hathor.manager import HathorManager
+from hathor.transaction import BaseTransaction
+from hathor.types import AddressB58
+from hathor.websocket.exception import InvalidAddress, InvalidXPub, LimitExceeded
+
+
+@dataclass(frozen=True, slots=True)
+class AddressItem:
+    index: int
+    address: AddressB58
+
+
+@dataclass(frozen=True, slots=True)
+class VertexItem:
+    vertex: BaseTransaction
+
+
+class ManualAddressSequencer(AsyncIterable[AddressItem]):
+    """An async iterable that yields addresses from a list. More addresses
+    can be added while the iterator is being consumed.
+    """
+
+    ADDRESS_SIZE = 34
+
+    def __init__(self) -> None:
+        self.max_pending_addresses_size: int = 5_000
+        self.pending_addresses: deque[AddressItem] = deque()
+        self.await_items: Deferred | None = None
+
+        # Flag to mark when all addresses have been received so the iterator
+        # can stop yielding after the pending list of addresses is empty.
+        self._stop = False
+
+    def _resume_iter(self) -> None:
+        """Resume yield addresses."""
+        if self.await_items is None:
+            return
+        if not self.await_items.called:
+            self.await_items.callback(None)
+
+    def add_addresses(self, addresses: list[AddressItem], last: bool) -> None:
+        """Add more addresses to be yielded. If `last` is true, the iterator
+        will stop when the pending list of items gets empty."""
+        if len(self.pending_addresses) + len(addresses) > self.max_pending_addresses_size:
+            raise LimitExceeded
+
+        # Validate addresses.
+        for item in addresses:
+            if len(item.address) != self.ADDRESS_SIZE:
+                raise InvalidAddress(item)
```

Comment:
issue(blocking): There are some addresses that are encoded with 35 chars.

If we really want to validate the address we can decode it and check that the length in bytes is 25.
Or actually validate the address (checksum, network byte validations) but I think that just validating the length is better since we don't add too much processing for each address.

Reply:
As discussed, this never happens because the version byte guarantees the base58 length will always be 34.